### PR TITLE
Add BIP44 Key Generation with HD Mint path

### PR DESCRIPTION
**Pull request by riordant**
# BIP44 Upgrade
The key generation process within zcoind uses [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki), ie. it creates keys using a Hierarchical Deterministic tree. This keypath is currently set to the default within Bitcoin Core, `m/0'/0'/0'`. [BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki) is a standard for specifying how various cryptocurrencies should be positioned within the HD tree, and is the standard by which hardware wallet devices such as Ledger and Trezor create addresses. 

This PR updates our key generation process to use a BIP44 path. It uses the `coinType` parameter for Zcoin as described in [SLIP-0044](https://github.com/satoshilabs/slips/blob/master/slip-0044.md). Our BIP44 path for normal address creation is now: 
`m/44'/136'/0'/0/<address_index>`. 
(As no functionality exists within zcoind to select any differing parameter for `account` or `change`, these have remained as the constants `0'` and `0` respectively. This does not affect compatibility with BIP44 supporting wallets).

This has full compatibility with keys generated along `m/0'/0'/0'` as only new keys generated will use the path.

# HD Mints
Secondly, this PR adds mint seed generation using this path. What differs is the `change` parameter: For mints this parameter is `2`. 

The BIP44 path for mint seed creation is:
`m/44'/136'/0'/2/<seed_index>`. 

The mint secret is thus generated as follows:
- A Keypair is generated at `<seed_index>` with 256-bit `privkey`
- A 512-bit mint seed (`mintMaster`) is generated via: `HMAC-SHA512(SHA256(<seed_index>), privkey)`
- The 256-bit ECDSA key (`ecdsaSecKey`) for each mint is `SHA256(Bottom 256 bits of mintMaster)`
- The 256-bit randomness seed (`seedRandomness`) for each mint is `SHA256(Top 256 bits of mintMaster)`

Then:
    Public key is generated from `ecdsaSecKey` in the usual way
    The `ScalarFromSeed` function in `secp256k1/src/cpp/Scalar.cpp` is used to generate valid randomness. 

Other changes:
    - Modifications to `uint256`, `arith_uint256` and `hash` to allow for 512 bit operations
    - Movement of `CZerocoinEntry(V3)` to a new file, `primitives/zerocoin.h` (compilation issues otherwise)
    - Updates to `Scalar.h`: addition of a function to generate randomness from a seed, and a parameter to the `hash` function to toggle if a `mod_p` operation should be performed after the hash (set to `true` by default)

### CODE REVIEW

**Comment by jivanyan** [comment]
Maybe we can consider using HMAC algorithms instead of simple hashing? I suggest using HMAC-SH512(seedMaster, count) instead of just SHA512(seedMaster | count).

**Comment by yura-pakhuchiy** [comment]
I suggest to use BIP32/BIP43 if possible. With pseudo-BIP44 path - `m/44'/136'/account'/2'/mint'`.
https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki
https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki

**Comment by riordant** [comment]
@jivanyan Ok I agree with this, considering also that the other BIP standards use HMAC. So this is what I have now:

- `seedMaster` generated (256 bits, using the random generator from the `Key` class)
- Next seed: HMAC-SHA512(seedMaster, count)
- ECDSA key: SHA256(Bottom 256 bits of seed)
- randomness seed: SHA256(Top 256 bits of seed)

**Comment by riordant** [comment]
@yura-pakhuchiy I'm currently looking at upgrading core to use our BIP44 path. I would like to keep it consistent with the official specification. In Electrum, the default path is `m/44'/136'/0'`. However, the BIP44 spec is `m / purpose' / coin_type' / account' / change / address_index`, ie. the Electrum path does not derive the `change` path, and starts creating addresses at the `account` level. How do you think we should proceed?

**Comment by yura-pakhuchiy** [comment]
@riordant Hm, electrum does use `change / address_index` part at least for BIP39 seeds, otherwise it would be incompatible with trezor and ledger. You can check with [this tool](https://iancoleman.io/bip39/). I have not checked for native electrum seeds, but it will be really strange if it treats them differently.

**Comment by yura-pakhuchiy** [comment]
BTW, consider also using BIP39 seeds in the Zcoin Core and/or new GUI. It is much easier to backup it, rather than lengthy xprv.

**Comment by riordant** [comment]
@yura-pakhuchiy Well the default path shown when setting up a HW wallet (`m/44'/136'/0'`) leaves out the last two value, `change`. I assume it just resolves to `0'` then.
On BIP39, yes I intend to look into it afterward.

**Comment by riordant** [comment]
This is the standard Im thinking now: 
BIP44: `m / purpose' / coin_type' / account' / change / <address_index>`
Mint Path: `m/ 44' / 136' / 1' / 0' / <seed_index>`
Keypair generated at `<seed_index>` with 256-bit `privkey`
512 mint seed is `HMAC-SHA512(<seed_index>, privkey)`

**Comment by yura-pakhuchiy** [comment]
You are probably assuming that account is always 0, but it is not. `m/44' /136' /1'` corresponds to the second Zcoin account on Trezor/Ledger. Account number in path for mints should correspond to the account number for wallet, so user can restore both of them. I think there 2 options to use:
1. `m/44'/136'/account'/2'/...(mint/seed/whatever)`
    `change` can be 0 or 1 according to BIPs, so we can use 2 for mints to avoid clashes. However maybe xpub derived we from `m/44'/136'/account'` or `m/44'/136'/account'/{0,1}` will leak information about mints derived from `m/44'/136'/account'/2'`. And since xpub is semi-public information it is probably better to use 2nd option:
2. `m/44'/136'/1000+account'/...(mint/seed/whatever)`
    High enough constant (1000 in this example) is added to account number to derive mint seeds.

**Comment by riordant** [comment]
I'm all for the 1st option. In the second case, I really don't like hard-setting some limit on the number of addresses before we should start to consider mints vs normal keys.
Could you explain in more detail how the first option could lead to leaked information? All that's being used is the private key (and index) from the key generated, I don't see how this could leak any info, assuming the same level of security as normal addresses.

**Comment by yura-pakhuchiy** [comment]
If you know `m/44'/136'/account'` you can generate `m/44'/136'/account'/...whatever`, eg. `m/44'/136'/account'/2'/0'`. Of course we do not tell anybody `m/44'/136'/account'`, only xpub generated from it, but even xpub leaks some information about `m/44'/136'/account'`, but I'm not cryptographer to estimate how bad it is. Also note that path components without hash are less secure than with hash. This is probably why Bitcoin Core have not adopted BIP44. Maybe @jivanyan can comment whether it is safe to use `m/44'/136'/account'/2'/mint_index'`?

**Comment by riordant** [comment]
Ok let's wait for @jivanyan to weigh in. If we don't feel it's secure I won't proceed with it

**Comment by jivanyan** [comment]
> If you know `m/44'/136'/account'` you can generate `m/44'/136'/account'/...whatever`, eg. `m/44'/136'/account'/2'/0'`. Of course we do not tell anybody `m/44'/136'/account'`, only xpub generated from it, but even xpub leaks some information about `m/44'/136'/account'`, but I'm not cryptographer to estimate how bad it is. Also note that path components without hash are less secure than with hash. This is probably why Bitcoin Core have not adopted BIP44. Maybe @jivanyan can comment whether it is safe to use `m/44'/136'/account'/2'/mint_index'`?


Having public keys generated from `m/44'/136'/account'/{0,1}' ` tell nothing about private keys generated from the path `m/44'/136'/account'/2''` so I think the first option is secure

**Comment by jivanyan** [comment]
> You are probably assuming that account is always 0, but it is not. `m/44'/136'/1'` corresponds to the second Zcoin account on Trezor/Ledger. Account number in path for mints should correspond to the account number for wallet, so user can restore both of them. I think there 2 options to use:
> 
> 1. `m/44'/136'/account'/2'/mint_index'`
>    `change` can be 0 or 1 according to BIPs, so we can use 2 for mints to avoid clashes. However maybe xpub derived we from `m/44'/136'/account'` or `m/44'/136'/account'/{0,1}` will leak information about mints derived from `m/44'/136'/account'/2'`. And since xpub is semi-public information it is quite dangerous.
> 2. `m/44'/136'/1000000+account'/mint_index'`
>    High enough constant (1000000 in this example) is added to account number to derive mint seeds. Nothing is leaked in this case, unless user will manually create account with such high number.
> 
> I think 1st is more close to BIP44, but since it may leak information, it is probably better to use 2nd option.


Having public keys generated from `m/44'/136'/account'/{0,1}' ` tell nothing about private keys generated from the path `m/44'/136'/account'/2''` so I think the first option is secure

**Comment by riordant** [comment]
Thank you for your input, @jivanyan. Given that I will proceed with the first option.

**Comment by yura-pakhuchiy** [comment]
@jivanyan Have you considered in your reply that:
1. Path for addresses is `m/44'/136'/account'/{0,1}`, not  `m/44'/136'/account'/{0,1}'`. Without hardening of the final component.
2. xpub is generated from  `m/44'/136'/account'`

**Comment by riordant** [comment]
@yura-pakhuchiy Do you happen to know why the `change` component is not hardened?

**Comment by yura-pakhuchiy** [comment]
IIRC for performance reasons. BIP44 was written by Trezor developers and Trezor has rather slow chip and limited memory inside.

**Comment by nopslide** [comment]
This pull request **introduces 38 alerts** when merging 036a125748e95b8e032287e54f98784e724451aa into 015fd012582324a1a90054718f7b3b067c95de49 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-57e434ce992fc4d9a10ec693787c723833780b32)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for Unused local variable
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 38 alerts** when merging 6a4a17f0c5a30a31ca5df48296dcb42980caecf0 into 015fd012582324a1a90054718f7b3b067c95de49 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-008128e8e7f67bdc0884b1d0e40c9a3262b284b6)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for Unused local variable
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 38 alerts** when merging e89954e5b5fd8be1b40ae1eb6b0ace70f46fa24e into 015fd012582324a1a90054718f7b3b067c95de49 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-cafc8ee1a846e9877c1735d003b2ea56110c4d9f)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for Unused local variable
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 38 alerts** when merging 6e791efdc9fcfcd9269648c5ae472e41e0be59c9 into 015fd012582324a1a90054718f7b3b067c95de49 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-5f05b621c89f40cf73141aaf5c05c37f560aea5f)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for Unused local variable
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 38 alerts** when merging bcb4ad0988af6f3900ea82d64dc82ca8e85754ae into 015fd012582324a1a90054718f7b3b067c95de49 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-44236533bbdb54e2b5a867bfa7b291b6684cec66)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for Unused local variable
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 38 alerts** when merging 6efbb34d2597a3b5c22e00f3005289d836cecd4c into 015fd012582324a1a90054718f7b3b067c95de49 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-ee0abe6f553ceeaf3daf86a0e34c271a4fc932c4)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for Unused local variable
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 38 alerts** when merging ae7bd96508f4772175d875764b8800f44ee3401d into 8a33b1378a6c91033025307d6d4b2a1f202894b8 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-70a0ad9e6c8228b48ce81bc367b18577c7b8ea45)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for Unused local variable
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jivanyan** [comment]
> @jivanyan Have you considered in your reply that:
> 
> 1. Path for addresses is `m/44'/136'/account'/{0,1}`, not  `m/44'/136'/account'/{0,1}'`. Without hardening of the final component.
> 2. xpub is generated from  `m/44'/136'/account'`
> 
> ?
I have read the BIP0032 again. In general, non-hardened keys will enable to generate the child public keys from the parent public key, but no information is revealed about the child private Keys. So even if xPub is public and the final components of  `m/44'/136'/account'/{0,1}` are not hardened, this does not violate the security of the mintMaster which is generated from the private keys.

**Comment by riordant** [comment]
@jivanyan When you say "final components" you are considering both `account` and `change` right? Because neither of these are hardened in the BIP44 spec

**Comment by nopslide** [comment]
This pull request **introduces 35 alerts** when merging b823489039f8ba009fe56a898f5f24b1d2c597a3 into 8a33b1378a6c91033025307d6d4b2a1f202894b8 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-a47a254a8d4de229a7f98ffe42e01b30ac808510)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Unused local variable
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 35 alerts** when merging d06564ef52f24dc8ec3e5c1fa092388b751f1cde into 8a33b1378a6c91033025307d6d4b2a1f202894b8 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-99fedfb08d278d749268819167e7f58117c5d7a1)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Unused local variable
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by ultimaweapon** [comment]
Is this one ready for review? I mean is specification already finalized?

**Comment by riordant** [comment]
The code in the PR is the spec as of @jivanyan 's latest comment. Perhaps should wait until he can reply, or if @yura-pakhuchiy has any more input

**Comment by nopslide** [comment]
This pull request **introduces 35 alerts** when merging 80e5183efb3a7a08340af9086eb645da3ac8d336 into 8a33b1378a6c91033025307d6d4b2a1f202894b8 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-f3346124310b49de12edfa394f03409b8cbbbff7)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Unused local variable
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by nopslide** [comment]
This pull request **introduces 35 alerts** when merging ad9b33cce70cb1c1555bffe03f68735fba038ad2 into 8a33b1378a6c91033025307d6d4b2a1f202894b8 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-3c56c1ede6eb315bd27e6998310eb04c04dc6c44)

**new alerts:**

* 26 for Duplicate key in dict literal
* 5 for &#39;import \*&#39; may pollute namespace
* 2 for Unused local variable
* 2 for Variable defined multiple times

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jivanyan** [comment]
> @jivanyan When you say "final components" you are considering both `account` and `change` right? Because neither of these are hardened in the BIP44 spec

Yes.  I consider both.

**Comment by thebevrishot** [comment]
@riordant please fix conflicts.

File: src/zerocoin_v3.h [hunk 1]
```diff
@@ -69,6 +69,15 @@ bool ConnectBlockSigma(
   const CBlock *pblock,
   bool fJustCheck=false);
 
+/*
+ * Get transaction hash from the chain using pubcoin value alone.
+ */
+bool ZerocoinGetSigmaMintTxHash(uint256& txHash, uint256 pubCoinValueHash);
+bool ZerocoinGetSigmaMintTxHash(uint256& txHash, GroupElement pubCoinValue);
```
Comment (levonpetrosyan93) [inline src/zerocoin_v3.h]: I think we need to remove Zerocoin form this function names.

File: src/zerocoin_v3.cpp [hunk 2]
```diff
@@ -754,10 +830,31 @@ bool CSigmaState::IsUsedCoinSerial(const Scalar &coinSerial) {
...
+    return false;
+}
+
 bool CSigmaState::HasCoin(const sigma::PublicCoin& pubCoin) {
     return mintedPubCoins.find(pubCoin) != mintedPubCoins.end();
 }
 
+bool CSigmaState::HasCoinHash(GroupElement &pubCoinValue, const uint256 &pubCoinValueHash) {
+    for ( auto it = mintedPubCoins.begin(); it != mintedPubCoins.end(); ++it ){
+    	sigma::PublicCoin pubCoin = (*it).first;
```
Comment (levonpetrosyan93) [inline src/zerocoin_v3.cpp]: Better to use const reference here. No need to copy.

File: src/wallet/walletdb.cpp [hunk 3]
```diff
@@ -1250,3 +1261,187 @@ bool CWalletDB::WriteHDChain(const CHDChain &chain) {
...
+    if (!WriteHDMint(dMint))
+        return error("%s: failed to write deterministic mint", __func__);
+
+    if (!Erase(make_pair(string("dzco"), dMint.GetPubCoinHash())))
+        return error("%s : failed to erase archived deterministic mint", __func__);
+
+    return true;
+}
+
+bool CWalletDB::UnarchiveZerocoinMint(const uint256& hashPubcoin, CSigmaEntry& zerocoin)
```
Comment (levonpetrosyan93) [inline src/wallet/walletdb.cpp]: I think it is better to use Sigma in place of Zerocoin.

File: src/zerocoin_v3.cpp (hunk 2 above)
Comment (riordant) [reply src/zerocoin_v3.cpp]: Do you mean `const` for `pubCoinValue`?

File: src/wallet/walletdb.cpp (hunk 3 above)
Comment (riordant) [reply src/wallet/walletdb.cpp]: I agree with this, however note that this is the case in the main `sigma` branch.

File: src/zerocoin_v3.cpp (hunk 2 above)
Comment (levonpetrosyan93) [reply src/zerocoin_v3.cpp]: You can just write const sigma::PublicCoin& pubCoin here..

Comment (riordant) [reply src/zerocoin_v3.cpp]: Sure

File: src/secp256k1/src/cpp/Scalar.cpp [hunk 4]
```diff
@@ -228,7 +245,8 @@ Scalar Scalar::hash(const unsigned char* data, size_t len) {
      throw "Scalar: hashing overflowed";
     }
     Scalar result_(&result);
-    result_.mod_p();
+    if(mod)
+        result_.mod_p();
```
Comment (thebevrishot) [inline src/secp256k1/src/cpp/Scalar.cpp]: why should not always mod?

File: src/arith_uint256.h [hunk 5]
```diff
@@ -295,7 +299,21 @@ class arith_uint256 : public base_uint<256> {
     friend arith_uint256 UintToArith256(const uint256 &);
 };
 
+/** 512-bit unsigned big integer. */
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/arith_uint256.h]: Do you think they should be separate?

File: src/hdmint/chain.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+class CBigNum;
+struct CMintMeta;
+class CTransaction;
+class CTxIn;
+class CTxOut;
+class CValidationState;
+class CZerocoinEntry;
+class uint256;
+
+bool IsSerialInBlockchain(const Scalar& bnSerial, int& nHeightTx);
```
Comment (levonpetrosyan93) [inline src/hdmint/chain.h]: Seems this function is never used.

File: src/hdmint/mintpool.cpp [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+bool CMintPool::Has(const CKeyID& seedId)
+{
+    return static_cast<bool>(count(seedId));
+}
+
+bool CMintPool::Get(const CKeyID& seedId, std::pair<CKeyID, uint32_t>& result)
+{
+    for(std::list<pair<CKeyID, uint32_t>>::iterator it = List().begin(); it != List().end(); ++it){
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/mintpool.cpp]: How do you think it should be iterated?

File: src/secp256k1/src/cpp/Scalar.cpp (hunk 4 above)
Comment (riordant) [reply src/secp256k1/src/cpp/Scalar.cpp]: As far as I could tell, it's inn-accurate to immediately take a hash outside of the range (0...p-1). So we call the hash function until we get a result within the correct range. Maybe this is unnecessary though. @jivanyan any input?

File: src/hdmint/chain.h (hunk 6 above)
Comment (riordant) [reply src/hdmint/chain.h]: Correct.

File: src/secp256k1/src/cpp/Scalar.cpp (hunk 4 above)
Comment (levonpetrosyan93) [reply src/secp256k1/src/cpp/Scalar.cpp]: Why we call serialize() and generate() in memberFromSeed? Actually we do the same thing 2 times.
I think we just need to call has() and always do mod_p() here.

File: src/arith_uint256.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (thebevrishot) [reply src/arith_uint256.h]: I think so.

File: src/hdmint/hdmint.cpp [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+void CHDMint::SetNull()
+{
+    nCount = 0;
+    seedId.SetNull();
+    hashSerial.SetNull();
+    txid.SetNull();
+    nHeight = -1;
+    nId = -1;
+    denom = 5; // SIGMA_ERROR
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/hdmint.cpp]: Agreed

File: src/hdmint/mintpool.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (thebevrishot) [reply src/hdmint/mintpool.cpp]: Sorry, I didn't describe clearly. The problem is the code call List() twice and them return different objects. And comparing iterator on different objects isn't safe. `it != List().end()`

I think it should be something like

```
auto mintList = List();
for(std::list<pair<CKeyID, uint32_t>>::iterator it = mintList.begin(); it != mintList.end(); ++it){
```

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -2086,11 +2113,23 @@ CAmount CWallet::GetDenominatedBalance(bool unconfirmed) const {
...
 {
     std::vector<CRecipient> vecSend;
+    CHDMint dMint;
+
+    uint32_t nCountLastUsed = zwalletMain->GetCount();
+
+
     std::transform(coins.begin(), coins.end(), std::back_inserter(vecSend),
-        [](const sigma::PrivateCoin& coin) -> CRecipient {
+        [&vDMints, &dMint](sigma::PrivateCoin& coin) -> CRecipient {
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/wallet.cpp]: In the `GenerateHDMint` function, coin is potentially modified. Does it need to be const?

File: src/wallet/wallet.cpp [hunk 10]
```diff
@@ -2187,9 +2231,15 @@ CAmount CWallet::SelectSpendCoinsForAmount(
 
 std::list<CSigmaEntry> CWallet::GetAvailableCoins() const {
     LOCK2(cs_main, cs_wallet);
-
+    CWalletDB walletdb(strWalletFile);
     std::list<CSigmaEntry> coins;
-    CWalletDB(strWalletFile).ListSigmaPubCoin(coins);
+    std::vector<CMintMeta> vecMists = pwalletMain->hdMintTracker->ListMints(true, true, true);
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/wallet.cpp]: Agreed

File: src/hdmint/mintpool.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/mintpool.cpp]: Agree.

File: src/hdmint/mintpool.cpp [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+            return true;
+        }
+    }
+    return false;
+}
+
+bool CMintPool::Get(const uint32_t& nCount, std::pair<CKeyID, uint32_t>& result)
+{
+    std::list<pair<CKeyID, uint32_t> > listMints = List();
+    for(std::list<pair<CKeyID, uint32_t>>::iterator it = listMints.begin(); it != listMints.end(); ++it){
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/mintpool.cpp]: Agree

File: src/hdmint/tracker.cpp [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+    meta.isUsed = dMint.IsUsed();
+    meta.hashSerial = dMint.GetSerialHash();
+    meta.denom = dMint.GetDenomination();
+    meta.isArchived = isArchived;
+    meta.isDeterministic = true;
+    if (! iszerocoinWalletInitialized)
+        zerocoinWallet = new CHDMintWallet(strWalletFile);
+    meta.isSeedCorrect = true;
+    if (! iszerocoinWalletInitialized)
+        delete zerocoinWallet;
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/tracker.cpp]: `zerocoinWallet` is initialized here: https://github.com/zcoinofficial/zcoin/blob/c8d46454c36f969d03756328ee810e7876301561/src/wallet/wallet.cpp#L7506

File: src/hdmint/chain.h [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+bool IsSerialInBlockchain(const uint256& hashSerial, int& nHeightTx, uint256& txidSpend);
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/chain.h]: Agreed

File: src/hdmint/chain.cpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+#include "txdb.h"
+#include "ui_interface.h"
+#include "zerocoin_v3.h"
+#include "wallet/wallet.h"
+#include "sigma/coin.h"
+
+// 6 comes from OPCODE (1) + vch.size() (1) + BIGNUM size (4)
+#define SCRIPT_OFFSET 6
+// For Script size (BIGNUM/Uint256 size)
+#define BIGNUM_SIZE   4
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/hdmint/chain.cpp]: Correct. Removed

File: src/hdmint/tracker.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (thebevrishot) [reply src/hdmint/tracker.cpp]: ok, then this is expected thing. will close this.

File: src/wallet/wallet.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (thebevrishot) [reply src/wallet/wallet.cpp]: first version don't write this object but ok if need to do.

File: src/arith_uint256.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (thebevrishot) [reply src/arith_uint256.h]: Ok, just leave it, this one isn't important.

File: src/qt/walletmodel.cpp [hunk 15]
```diff
@@ -968,10 +968,11 @@ void WalletModel::sigmaMint(const CAmount& n)
             return sigma::PrivateCoin(sigmaParams, denom);
         });
 
-    auto recipients = CWallet::CreateSigmaMintRecipients(privCoins);
+    vector<CHDMint> vDMints;
+    auto recipients = CWallet::CreateSigmaMintRecipients(privCoins, vDMints);
 
     CWalletTx wtx;
-    std::string strError = pwalletMain->MintAndStoreSigma(recipients, privCoins, wtx);
+    std::string strError = pwalletMain->MintAndStoreSigma(recipients, privCoins, vDMints, wtx);
```
Comment (thebevrishot) [inline src/qt/walletmodel.cpp]: not here too but please add same thing to function `updateSigmaCoins`. Actually better to use function that get all coins(both HD and non-HD mints) but I don't know the function.

File: src/wallet/wallet.cpp [hunk 16]
```diff
@@ -6099,35 +6185,23 @@ string CWallet::MintAndStoreSigma(const vector<CRecipient>& vecSend,
...
-        sigma::PublicCoin checkPubCoin(zerocoinTx.value, privCoin.getPublicCoin().getDenomination());
-        if (!checkPubCoin.validate()) {
-            return "error: pubCoin not validated.";
-        }
-        zerocoinTx.randomness = privCoin.getRandomness();
-        zerocoinTx.serialNumber = privCoin.getSerialNumber();
-        const unsigned char *ecdsaSecretKey = privCoin.getEcdsaSeckey();
-        zerocoinTx.ecdsaSecretKey = std::vector<unsigned char>(ecdsaSecretKey, ecdsaSecretKey+32);
-        walletdb.WriteZerocoinEntry(zerocoinTx);
-        NotifyZerocoinChanged(this,
```
[the remark this answers is not in the document]
Comment (riordant) [reply src/wallet/wallet.cpp]: Agreed

File: src/qt/walletmodel.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (riordant) [reply src/qt/walletmodel.cpp]: Perfect. Agreed

Comment (riordant) [reply src/qt/walletmodel.cpp]: Agreed. Added to both

**Comment by ultimaweapon** [comment]
@riordant please fix CI failure.

**Pull request merged**